c10/core/PythonDispatcher.h

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <memory>
+
+#include <c10/core/TensorImpl.h>
+#include <c10/macros/Macros.h>
+#include <c10/util/python_stub.h>
+
+namespace c10 {
+
+// NOTE [What is PythonDispatcher?]
+// A PythonDispatcher represents the type of a Tensor subclass that has
+// a __torch_dispatch__ classmethod. Concretely, it holds the class as a
+// PyObject* and a PyInterpreter* that says which python interpreter the
+// class came from.
+//
+// See NOTE [dispatch_fn's type argument] for more details
+struct C10_API PythonDispatcher {
+  explicit PythonDispatcher(PyObject* type,
+                            c10::impl::PyInterpreter* interpreter) noexcept;
```

Comment:
We should add a comment that PythonDispatcher steals the reference to the object to avoid confusion when folks use it